cpp/src/arrow/dataset/filter.cc

```diff
@@ -772,13 +772,10 @@ std::shared_ptr<Expression> and_(std::shared_ptr<Expression> lhs,
 }
 
 std::shared_ptr<Expression> and_(const ExpressionVector& subexpressions) {
-  if (subexpressions.size() == 0) {
-    return scalar(true);
-  }
   return std::accumulate(
-      subexpressions.begin(), subexpressions.end(), std::shared_ptr<Expression>(),
+      subexpressions.begin(), subexpressions.end(), scalar(true),
       [](std::shared_ptr<Expression> acc, const std::shared_ptr<Expression>& next) {
-        return acc == nullptr ? next : and_(std::move(acc), next);
+        return acc->Equals(true) ? next : and_(std::move(acc), next);
```

Comment:
Hmm... As a sidenote, I'm sure this would be more readable (at least to me!) if rewritten as a plain loop rather than a `std::accumulate` call.